sw/device/tests/dif/dif_plic_unittest.cc

```diff
@@ -68,7 +68,24 @@ TEST_F(InitTest, Success) {
 
 class IrqTest : public PlicTest {
  protected:
-  IrqTest() { EXPECT_EQ(RV_PLIC_PARAM_NUMTARGET, 1); }
+  IrqTest() {
+    // Make sure to change the `last_bit` when `RV_PLIC_PARAM_NUMSRC` changes.
+    // As `last_bit` represents the bit index in a register, we need to count
+    // all of the last bits of a multireg to get the total number of bits.
+    // The bit count in IE, LE and IP registers is expected to be the same.
+    //
+    // This check has been added to help diagnose the mismatch of test values
+    // with the HW defines. One of the recent PRs ran into this problem, and
+    // the failure message was not descriptive, so some engineering time was
+    // lost to investigation.
+    uint8_t number_of_sources = 0;
+    for (const auto &reg : kEnableRegisters) {
+      number_of_sources += (reg.last_bit + 1);
+    }
+    EXPECT_EQ(RV_PLIC_PARAM_NUMSRC, number_of_sources);
```

Comment:
@silvestrst as a follow-up, it seems [one can add diagnostics when the expect fails](https://stackoverflow.com/a/16491892/261393), here it might be an idea to be more explicit that they need to update `kEnableRegisters` and the like when RV_PLIC_PARAM_NUMSRC changes, given the feedback that this was still hard for Martin to debug.

Reply:
@lenary that`s a great point - I didn't know about that feature. Thanks!

Reply:
Is there written guidance on how to accurately update kEnableRegisters?
With I2C, we've crossed into over 100 registers, and now the edit that I've made for pattgen doesn't work.  It looks like there is something about crossing a 96 bit boundary?

Reply:
For instance my tests are now failing on this branch: 
https://github.com/martin-lueker/opentitan/tree/i2c_eg

Any suggestions?

Reply:
Oh, ick. It seems the plic only supports 96 interrupts.  I'll have to figure out what to do about this.

Reply:
Never mind.  It turns out that I'm looking at the online docs, not my locally generated ones.  Sorry for the confusion.

Reply:
No problem, but in general the way I have implemented this unittest - `kEnableRegisters` and other registers have an `offset` and `last_bit` fields. Which then are used to loop through all of the bits in the register. The unittest uses `kEnableRegisters` and similar for setting test expectations, and uses `RV_PLIC_PARAM_NUMSRC` for function calls.

```
  static constexpr std::array<Register, RV_PLIC_IE0_MULTIREG_COUNT>
      kEnableRegisters{{
          {RV_PLIC_IE0_0_REG_OFFSET, RV_PLIC_IE0_0_E_31_BIT},
          {RV_PLIC_IE0_1_REG_OFFSET, RV_PLIC_IE0_1_E_63_BIT},
          {RV_PLIC_IE0_2_REG_OFFSET, RV_PLIC_IE0_2_E_87_BIT},
      }};
```
We have 3 entries because `IE` is multireg and X interrupt sources are spread across Y registers, with the last one potentially having less than 32 interrupt sources.

Reply:
Hi @silvestrst, this helps!
  Specifically though, the problem I just had is that my interrupt count just created a fourth entry, and this was causing me confusion.  There are a number of places that need to be updated in this case:  In this assignment, the declaration (I'm fixing to use the register count macros instead of fixed values), and in `ExpectInitReset`.   So after this PR there will be two places in this file that need to be changed if the interrupt counts need to be changed.

  Documentation doesn't hurt, but my confusion mainly started with the fact that the PLIC is autogenerated, and this code is static.  So in this situation, peripheral designers (like me) have tons of infrastructure to help us integrate this into top level, and then we get to this CI test, and there is a lot of initial confusion. We don't know where to start!  Even the _online documentation_ is out of date, because some IP's like this PLIC have been changed (but the documentation is based on master.)   It can all be figured out and documented, but autogeneration of this test (perhaps from a template) would be even better and more inline with the flow on the HW side.  Is autogeneration of this test feasible given our build flow?

Reply:
If more documentation is added, and the DIF tests require some changes every time IP's are added or removed, then perhaps we can make a "SW Steps" document, and the key part here would be to link it at the top of the [topgen.py](https://docs.opentitan.org/doc/rm/topgen_tool/) documentation.   Making docs easy to find, and cross referenced is just as important as writing them.

Reply:
@martin-lueker I made the decision that this version of the PLIC should effectively be hard-coded in this manner. We do want to move to a place where DIFs (especially those affected by topgen) are parameterised in a manner that makes IP easier to integrate, but this is a lower priority than completing e.g. the Mask ROM.  For an example of a DIF that was written from the start to be parameterisable in the way we want, see `dif_rv_timer.h`.

I am also keen that the tests do not compute the highest register/bit offsets in the same way the DIF implementation does, so we are actually properly testing the mapping computed in the DIF.

I realise there's a bit of a bump to integrating an IP with making the build pass (and you had a bigger bump this time as you overflowed the existing registers), but I think that's better than the alternative of the build passing but the software failing on the hardware for unknown and hard-to-debug reasons. This is why I'm keen we favour adding documentation to ensure it's easier to rectify what's going wrong in these tests.

I hope this is understandable? If you or your team run into this again in future, please do send us a message in the software room in slack, and someone will be able to help you find the places that need to be changed. This applies not just for the PLIC, but also for other DIFs or failing software builds.